Gems/Atom/Feature/Common/Code/Source/RayTracing/RayTracingRingBuffer.h

```diff
@@ -0,0 +1,92 @@
+#include <Atom/RPI.Public/Buffer/Buffer.h>
+#include <Atom/RPI.Public/Buffer/BufferSystemInterface.h>
+#include <AzCore/std/containers/span.h>
+#include <AzCore/std/string/string.h>
+
+namespace AZ::Render
+{
+    //! A class which manages a FrameCountMax number of RPI buffers and manages them in a ring buffer structure, meaning that whenever data
+    //! needs to be updated, the current buffer index is incremented (mod FrameCountMax) and the data is then written to the new current
+    //! buffer, such that the other buffers stay valid.
+    class RayTracingRingBuffer : public RHI::FrameCountMaxRingBuffer<Data::Instance<RPI::Buffer>>
+    {
+        AZStd::string m_bufferName;
+        RPI::CommonBufferPoolType m_bufferPoolType;
+        uint32_t m_elementSize;
+        RHI::Format m_bufferFormat{ RHI::Format::Unknown };
+
+    public:
+        //! Creates a set of buffers with a name and a format. The element size is derived from the format.
+        RayTracingRingBuffer(const AZStd::string& bufferName, RPI::CommonBufferPoolType bufferPoolType, RHI::Format bufferFormat);
+
+        //! Creates a set of buffers with a name and an element size. The format is set to unknown.
+        RayTracingRingBuffer(const AZStd::string& bufferName, RPI::CommonBufferPoolType bufferPoolType, uint32_t elementSize);
+
+        //! Returns true if the current buffer was already created and is not empty.
+        bool IsCurrentBufferValid() const;
+
+        //! Returns the current buffer
+        const Data::Instance<RPI::Buffer>& GetCurrentBuffer() const;
+
+        //! Returns a RHI view of the current buffer
+        const RHI::BufferView* GetCurrentBufferView() const;
+
+        //! Increments the current buffer index, potentially resized the current buffer and updates the data of the current data. This is a
+        //! convenience function which calls AdvanceCurrentBuffer(), EnsureCurrentBufferSizeAtLeast(...) and UpdateCurrentBufferData(...)
+        void AdvanceCurrentBufferAndUpdateData(const void* data, u64 dataSizeInBytes);
+
+        //! Increments the current buffer index, potentially resized the current buffer and updates the data of the current data. This is a
+        //! convenience function which calls AdvanceCurrentBuffer(), EnsureCurrentBufferSizeAtLeast(...) and UpdateCurrentBufferData(...)
+        template<typename T>
+        void AdvanceCurrentBufferAndUpdateData(AZStd::span<const T> data)
+        {
+            AdvanceCurrentBufferAndUpdateData(data.data(), data.size() * sizeof(T));
+        }
+
+        //! Convenience function which allows automatic conversion from vector/array to span
+        template<typename T>
+        void AdvanceCurrentBufferAndUpdateData(const T& data)
+        {
+            AdvanceCurrentBufferAndUpdateData<typename T::value_type>(data);
+        }
+
+        //! Creates or resizes the current buffer to fit the given number of bytes.
+        void EnsureCurrentBufferSizeAtLeast(u64 bufferSizeInBytes);
```

Comment:
Why not just simply call this function CreateOrResizeBuffer? To me EnsureCurrentBufferSizeAtLeast is confusing.